Implement enforcement of `require-trusted-types-for` CSP directive

lukewarlow opened the pull request:
#### caf4cd7788f1f9bce326f9c5c145ff04e32dad07
<pre>
Implement enforcement of `require-trusted-types-for` CSP directive
<a href="https://bugs.webkit.org/show_bug.cgi?id=267685">https://bugs.webkit.org/show_bug.cgi?id=267685</a>

Reviewed by Darin Adler.

This patch implements the StringContext idl attribute to check
the `require-trusted-types-for` CSP and enforce trusted types accordingly.

This patch also makes use of the StringContext IDL attribute on an initial set of sinks.

More complicated sinks such as setAttribute, execCommand, eval and timer functions will be addressed in follow ups.

Spec: <a href="https://w3c.github.io/trusted-types/dist/spec/#integrations">https://w3c.github.io/trusted-types/dist/spec/#integrations</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/reporting/report-clips-sample.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-source-file-path-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::Converter&lt;IDLStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedHTMLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedScriptAdaptor&lt;T&gt;&gt;::convert):
(WebCore::Converter&lt;IDLAtomStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
(WebCore::JSConverter&lt;IDLAtomStringStringContextTrustedScriptURLAdaptor&lt;T&gt;&gt;::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateParametersCheck):
(IsAnnotatedType):
(GetAnnotatedIDLType):
(JSValueToNative):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestStringContextConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.cpp: Added.
(WebCore::JSTestStringContextDOMConstructor::prototypeForStructure):
(WebCore::JSTestStringContextDOMConstructor::initializeProperties):
(WebCore::JSTestStringContextPrototype::finishCreation):
(WebCore::JSTestStringContext::JSTestStringContext):
(WebCore::JSTestStringContext::createPrototype):
(WebCore::JSTestStringContext::prototype):
(WebCore::JSTestStringContext::getConstructor):
(WebCore::JSTestStringContext::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringBody):
(WebCore::JSTestStringContext::subspaceForImpl):
(WebCore::JSTestStringContext::analyzeHeap):
(WebCore::JSTestStringContextOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestStringContextOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestStringContext::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.h: Added.
(WebCore::JSTestStringContext::create):
(WebCore::JSTestStringContext::createStructure):
(WebCore::JSTestStringContext::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestStringContext.idl: Added.
* Source/WebCore/dom/Document+HTML.idl:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element+DOMParsing.idl:
* Source/WebCore/dom/InnerHTML.idl:
* Source/WebCore/dom/Range+DOMParsing.idl:
* Source/WebCore/dom/TrustedType.cpp: Added.
(WebCore::TrustedTypeVisitor::operator()):
(WebCore::trustedTypeToString):
(WebCore::trustedTypeToCallbackName):
(WebCore::processValueWithDefaultPolicy):
(WebCore::trustedTypeCompliantString):
* Source/WebCore/dom/TrustedType.h: Copied from Source/WebCore/workers/Worker.idl.
* Source/WebCore/dom/TrustedTypePolicy.cpp:
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
(WebCore::TrustedTypePolicy::getPolicyValue):
* Source/WebCore/dom/TrustedTypePolicy.h:
* Source/WebCore/html/HTMLEmbedElement.idl:
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLObjectElement.idl:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
(WebCore::ContentSecurityPolicy::allowMissingTrustedTypesForSinkGroup const):
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::shouldReportSample const):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
(WebCore::ContentSecurityPolicyDirectiveList::requiresTrustedTypesForScript const):
* Source/WebCore/workers/Worker.idl:
* Source/WebCore/workers/WorkerGlobalScope.idl:
* Source/WebCore/workers/service/ServiceWorkerContainer.idl:
* Source/WebCore/workers/shared/SharedWorker.idl:
* Source/WebCore/xml/DOMParser.idl:

Canonical link: <a href="https://commits.webkit.org/275607@main">https://commits.webkit.org/275607@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f762541d6bc6a352e312cab66254349ee1c8457

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42279 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44673 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44877 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38396 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44586 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35033 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42853 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18234 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36416 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15964 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37456 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46334 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38477 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37785 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41696 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14086 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40286 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18714 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9462 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18776 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18359 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/JavaScriptCore/runtime/Error.h:
```diff
@@ -10,7 +10,7 @@
  *  This library is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty of
- *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNthrowTypeErrorU
```
TODO: revert

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +127,352 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
 
+String valueToTrustedHTMLString(JSGlobalObject& lexicalGlobalObject, JSValue value, const String& sink, bool nullToEmpty)
+{
+    // TODO: check settings for feature flag in all these functions
```
TODO add a settings feature flag check in all these functions

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +127,352 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    // TODO: check settings for feature flag in all these functions
+    VM& vm = lexicalGlobalObject.vm();
+
+    if (auto* trustedHTML = JSTrustedHTML::toWrapped(vm, value))
```
TODO extract shared code into common functions to avoid so much code duplication. Probably best to follow the algorithm breakdown within the spec for optimal reusability.

lukewarlow commented on Source/WebCore/html/HTMLScriptElement.idl:
```diff
@@ -22,7 +22,6 @@
 ] interface HTMLScriptElement : HTMLElement {
     [CEReactions=NotNeeded] attribute ScriptString text;
     [CEReactions=Needed] attribute [LegacyNullToEmptyString] ScriptString innerText;
-    [CEReactions=Needed] attribute ScriptString? textContent;
```
This change should be done inside of #22909

lukewarlow commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -120,12 +120,12 @@ TrustedTypePolicyFactory* WorkerGlobalScopeTrustedTypes::trustedTypes() const
 // WindowOrWorkerGlobalScopeTrustedTypes.
 
-ExceptionOr<TrustedTypePolicyFactory*> WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(ScriptExecutionContext&, LocalDOMWindow& window)
+TrustedTypePolicyFactory* WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(LocalDOMWindow& window)
```
These changes have been made in #22303 so once rebased they will go

lukewarlow commented on Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
```diff
@@ -132,7 +133,7 @@ class ContentSecurityPolicyDirectiveList {
     bool m_upgradeInsecureRequests { false };
     bool m_hasBlockAllMixedContentDirective { false };
-    bool m_requireTrustedTypes { false };
+    bool m_requireTrustedTypesForScript { false };
```
Make this name change in #22822

lukewarlow commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +147,126 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+            auto required = !contentSecurityPolicy->allowMissingTrustedTypesForSinkGroup(expectedType, sink, "script"_s, stringValue);
+
+            if (required)
+                throwTypeError(*scriptExecutionContext.globalObject(), scope, makeString("This assignment requires a ", expectedType));
```
```suggestion
                throwTypeError(scriptExecutionContext.globalObject(), scope, makeString("This assignment requires a ", expectedType));
```

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +122,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+        ASSERT_NOT_REACHED();
+        return nullString();
+    }
+    auto scriptExecutionContext = jsDynamicCast<JSDOMGlobalObject*>(&lexicalGlobalObject)->scriptExecutionContext();
```
RefPtr

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +122,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
     return replaceUnpairedSurrogatesWithReplacementCharacter(WTFMove(string));
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-compliant-string-algorithm
```
d better if the function below would match the referenced algorithm line by line.
It could also be useful to reuse the same variable names (input, convertedInput...).

lukewarlow replied:
This is a slightly tricky one because the algorithm as specced is split across two functions within WebKit. One here which is largely working with IDL types and then the one in TrustedTypeUtils which is where most of the actual algorithm is implemented.

So it's not really possible to follow it line by line. I have however renamed value to input and lexicalGlobalObject to global.

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +122,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+    if (stringValue.isNull() && shouldConvertNullToEmptyString == ShouldConvertNullToEmptyString::Yes)
+        return emptyString();
+
+    return stringValue;
```
I do not see where we throw TypeError in this implementation. Is it expected, should we have a FIXME?

lukewarlow replied:
We throw it inside of TrustedTypesUtils::getTrustedTypeCompliantString() and it propogates up.

youennf replied:
It would be better for the DOM version of getTrustedTypeCompliantString to return an error and to throw here I think.

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -170,6 +175,225 @@ template<typename T> struct JSConverter<IDLLegacyNullToEmptyStringAdaptor<T>> {
+
+    static String convert(JSC::JSGlobalObject& lexicalGlobalObject, JSC::JSValue value, const String& sink)
+    {
+        return getTrustedTypeCompliantString("TrustedHTML"_s, lexicalGlobalObject, value, sink, ShouldConvertNullToEmptyString::Yes);
```
Would it make sense to introduce an enum instead of passing a String for TrustedHTML et al?

lukewarlow replied:
Yeah that's a good idea. Will do that.

youennf commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +50,74 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
-        if (callbackResult.type() == CallbackResultType::ExceptionThrown)
-            return Exception { ExceptionCode::ExistingExceptionError };
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-policy-value-algorithm
+ExceptionOr<String> TrustedTypePolicy::getPolicyValue(const String& trustedTypeName, const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&& arguments, bool throwIfMissing)
```
Use an enum for trustedTypeName as well?

youennf commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +50,74 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+
+    String result(nullString());
+
+    return result;
```
one liner? `return String(nullString())`?

youennf commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +50,74 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-policy-value-algorithm
+ExceptionOr<String> TrustedTypePolicy::getPolicyValue(const String& trustedTypeName, const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&& arguments, bool throwIfMissing)
+{
+    auto callbackName = emptyString();
```
Would read better if we were following line by line the spec (aka, get the function, throw if null, and call the function otherwise).

lukewarlow replied:
Because the 3 callbacks are all of different types I don't know how to do this without using a variant which would make the code needlessly complicated imo?

youennf commented on Source/WebCore/dom/Document+HTML.idl:
```diff
@@ -61,8 +61,8 @@ partial interface Document {
-    [CEReactions=Needed, CallWith=EntryDocument] undefined write(DOMString... text);
-    [CEReactions=Needed, CallWith=EntryDocument] undefined writeln(DOMString... text);
+    [CEReactions=Needed, CallWith=EntryDocument] undefined write(HTMLString... text);
+    [CEReactions=Needed, CallWith=EntryDocument] undefined writeln(HTMLString... text);
```
Can this create web compat issues?
Usually for those kind of features, we would introduce a feature flag (probably TrustedTypesEnabled here).
If flag is off by default, there would be no change compared to before.
Is this patch guaranteeing this?

lukewarlow replied:
I will double check the logic of the code but the StringContext attribute as implemented should be identical to it not being there when the setting is off.

youennf commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,41 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+bool ContentSecurityPolicy::requireTrustedTypesForSinkGroup(const String& sinkGroup) const
+{
+    bool required = false;
+    for (auto& policy : m_policies) {
```
anyOf?

youennf commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,41 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+
+bool ContentSecurityPolicy::allowMissingTrustedTypesForSinkGroup(const String& stringContext, const String& sink, const String& sinkGroup, const String& source) const
+{
+    String sourceURL;
```
sourceURL is only used in case of reportViolation, I would create close to where it is used.

youennf commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,41 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+
+            const ContentSecurityPolicyDirectiveList& list = *policy.get();
+
+            String sample = makeString(sink, "|"_s, StringView(source).left(40).toString());
```
Why 40?

lukewarlow replied:
That's how it's specced. Step 2.7 here https://w3c.github.io/trusted-types/dist/spec/#should-block-sink-type-mismatch

youennf replied:
Can we add a link to the algorithm just before ContentSecurityPolicy::allowMissingTrustedTypesForSinkGroup?

youennf replied:
Also is the toString() call needed?

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,147 @@
+
+    auto requireTrustedTypes = contentSecurityPolicy
+    ? contentSecurityPolicy->requireTrustedTypesForSinkGroup("script"_s)
+    : false;
```
indent

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,147 @@
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (!std::holds_alternative<std::nullptr_t>(convertedType)) {
+            auto dataString = std::visit(TrustedTypeVisitor { }, convertedType);
```
```suggestion
            stringValue = std::visit(TrustedTypeVisitor { }, convertedType);
```

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
 }
 
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-compliant-string-algorithm
+String getTrustedTypeCompliantString(const TrustedType expectedType, JSGlobalObject& global, JSValue input, const String& sink, ShouldConvertNullToEmptyString shouldConvertNullToEmptyString)
```
```suggestion
String getTrustedTypeCompliantString(TrustedType expectedType, JSGlobalObject& global, JSValue input, const String& sink, ShouldConvertNullToEmptyString shouldConvertNullToEmptyString)
```

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+{
+    VM& vm = global.vm();
+
+    if (expectedType == TrustedType::TrustedHTML) {
```
Let's use switch

youennf commented on Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
```diff
@@ -118,4 +121,39 @@ AtomString valueToUSVAtomString(JSGlobalObject& lexicalGlobalObject, JSValue val
+        ASSERT_NOT_REACHED();
+        return nullString();
+    }
+    RefPtr scriptExecutionContext = jsDynamicCast<JSDOMGlobalObject*>(&global)->scriptExecutionContext();
```
We might need to check for scriptExecutionContext.

lukewarlow replied:
Not sure what the best thing to do would be in the case where it doesn't exist? Currently I've got an ASSERT_NOT_REACHED() which doesn't neccessarily feel right?

youennf commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +51,72 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+    auto callbackName = emptyString();
+    CallbackResult<String> policyValue(CallbackResultType::UnableToExecute);
+    if (trustedTypeName == TrustedType::TrustedHTML) {
+        callbackName = "createHTML"_s;
```
callbackName is only used in the exception case, I would tend to get the callbackName from the enum in the throwIfMissing code path using trustedTypeToString.

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,170 @@
+    VM& vm = scriptExecutionContext.vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    RefPtr<TrustedTypePolicy> protector = nullptr;
```
```suggestion
    RefPtr<TrustedTypePolicy> protectedPolicy;
```

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,170 @@
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#process-value-with-a-default-policy-algorithm
+std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> processValueWithDefaultPolicy(ScriptExecutionContext& scriptExecutionContext, const TrustedType expectedType, const String& input, const String& sink)
```
```suggestion
std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> processValueWithDefaultPolicy(ScriptExecutionContext& scriptExecutionContext, TrustedType expectedType, const String& input, const String& sink)
```

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,170 @@
+    if (!requireTrustedTypes)
+        return stringValue;
+
+    std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> convertedInput = processValueWithDefaultPolicy(scriptExecutionContext, expectedType, stringValue, sink);
```
```suggestion
    auto convertedInput = processValueWithDefaultPolicy(scriptExecutionContext, expectedType, stringValue, sink);
```

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,170 @@
+    std::variant<std::nullptr_t, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> result;
+    if (expectedType == TrustedType::TrustedHTML) {
+        result.emplace<1>(TrustedHTML::create(policyValue));
+        return result;
```
Why do we need this dance here?
Can we simply do: `return { TrustedHTML::create(policyValue) }` or something?

youennf commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,170 @@
+    FixedVector<JSC::Strong<JSC::Unknown>> arguments({ strongExpectedType, strongSink });
+    auto policyValueHolder = protector->getPolicyValue(expectedType, input, WTFMove(arguments), false);
+    if (policyValueHolder.hasException()) {
+        propagateException(*scriptExecutionContext.globalObject(), scope, policyValueHolder.releaseException());
```
I would tend to return the exception here and let the caller deal with throwing.
Hopefully that would be the `getTrustedTypeCompliantString` with global object version.

lukewarlow replied:
I've rejigged this to hopefully work more how you'd like. I'm not sure how best to "throw" would be inside of JSDOMConvert, I've currently just called propagateException but in the other file instead?

youennf reviewed (commented):
Looks almost ready to me.

darinadler commented on LayoutTests/TestExpectations:
```diff
@@ -5467,14 +5467,18 @@ webkit.org/b/261849 imported/w3c/web-platform-tests/css/css-scroll-anchoring/heu
 webkit.org/b/261849 imported/w3c/web-platform-tests/css/css-scroll-anchoring/start-edge-in-block-layout-direction.html [ Skip ]
 
-# Trusted Types aren't implemented yet
+# Trusted Types aren't fully implemented yet
```
Why is it best to skip these tests instead of expecting failures? Normally we would only want to skip a test if it hangs or runs very slowly. It’s much better to just expect a failure.

lukewarlow replied:
I think all of these are timeouts because they timeout in the failure case.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -28,16 +28,22 @@
 #include "IDLTypes.h"
 #include "JSDOMConvertBase.h"
 #include "StringAdaptors.h"
+#include "TrustedTypeUtils.h"
```
We don’t normally name headers in WebKit `XXXUtils.h`. Seems this could just be called `TrustedType.h`?

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.h:
```diff
@@ -0,0 +1,50 @@
+    TrustedScriptURL,
+};
+
+String trustedTypeToString(const TrustedType);
```
No need to use `const TrustedType` here, just `TrustedType`.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    {
+        return nullString();
+    }
+    String operator()(Ref<TrustedHTML> value)
```
Can this be `const Ref<TrustedHTML>&` instead? Would save reference count churn.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    {
+        return value->toString();
+    }
+    String operator()(Ref<TrustedScript> value)
```
Can this be `const Ref<TrustedScript>&` instead? Would save reference count churn.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    {
+        return value->toString();
+    }
+    String operator()(Ref<TrustedScriptURL> value)
```
Can this be `const Ref<TrustedScriptURL>&` instead? Would save reference count churn.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    }
+};
+
+String trustedTypeToString(const TrustedType trustedType)
```
Return type should be `ASCIILiteral` instead of `String`, in some cases this will save creating and then destroying a temporary string.

Argument should be `TrustedType`, not `const TrustedType`.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+        return "TrustedScript"_s;
+    case TrustedType::TrustedScriptURL:
+        return "TrustedScriptURL"_s;
+    default:
```
Should leave out the `default` here so the compiler checks at compile time that we didn’t forget any of the enumeration values. Can still handle the fall-through case with code after the switch statement if you need to.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+{
+    RefPtr<TrustedTypePolicy> protectedPolicy;
+    if (RefPtr document = dynamicDowncast<Document>(scriptExecutionContext)) {
+        if (auto window = document->domWindow()) {
```
Perhaps the type here needs to be `RefPtr` instead of `auto`?

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    if (RefPtr document = dynamicDowncast<Document>(scriptExecutionContext)) {
+        if (auto window = document->domWindow()) {
+            auto trustedTypesFactory = WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(*window);
+            protectedPolicy = trustedTypesFactory->defaultPolicy();
```
Can we do this on one line instead?

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+        }
+    } else if (RefPtr workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(scriptExecutionContext)) {
+        auto trustedTypesFactory = WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(*workerGlobalScope);
+        protectedPolicy = trustedTypesFactory->defaultPolicy();
```
Can we do this on one line instead?

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    JSC::Strong<JSC::Unknown> strongExpectedType(vm, jsExpectedType);
+    auto jsSink = JSC::jsString(vm, sink);
+    JSC::Strong<JSC::Unknown> strongSink(vm, jsSink);
+    FixedVector<JSC::Strong<JSC::Unknown>> arguments({ strongExpectedType, strongSink });
```
I think we can write this:
```
FixedVector<JSC::Strong<JSC::Unknown>> arguments({ { vm, jsExpectedType }, { vm, jsSink } });
```
And get rid of the `strongExpectedType` and `strongSink` local variables.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+
+    auto requireTrustedTypes = contentSecurityPolicy
+        ? contentSecurityPolicy->requireTrustedTypesForSinkGroup("script"_s)
+        : false;
```
I would write this with `&&` instead of `? :`.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    String stringValue(input);
+
+    if (!scriptExecutionContext.settingsValues().trustedTypesEnabled)
+        return stringValue;
```
Might want `WTFMove(stringValue)` for this and the many other cases here.

lukewarlow replied:
WTFMove here causes build failures on the linux pipelines (WPE and GTK).

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+        auto allowMissingTrustedTypes = contentSecurityPolicy->allowMissingTrustedTypesForSinkGroup(trustedTypeToString(expectedType), sink, "script"_s, stringValue);
+
+        if (!allowMissingTrustedTypes)
+            return Exception { ExceptionCode::TypeError, makeString("This assignment requires a ", trustedTypeToString(expectedType)) };
```
Seems like we might want the message here to be more human readable.
> This assignment requires a trusted HTML string.

rather than

> This assignment requires a TrustedHTML

lukewarlow replied:
I have a follow up issue to go through the error messages to ensure they're as helpful as possible there's some rather specific cases where we can be more descriptive. I'm happy to address that then, or can do it now if you'd prefer.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.h:
```diff
@@ -0,0 +1,50 @@
+
+String trustedTypeToString(const TrustedType);
+
+WEBCORE_EXPORT std::variant<std::nullptr_t, Exception, Ref<TrustedHTML>, Ref<TrustedScript>, Ref<TrustedScriptURL>> processValueWithDefaultPolicy(ScriptExecutionContext&, TrustedType, const String& input, const String& sink);
```
Normally we’d use `std::monostate` instead of `std::nullptr_t` for this.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,42 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+            String sourceURL;
+            TextPosition sourcePosition(OrdinalNumber::beforeFirst(), OrdinalNumber());
+            String sample = makeString(sink, "|"_s, StringView(source).left(40));
+            reportViolation("require-trusted-types-for"_s, list, "trusted-types-sink"_s, consoleMessage, sourceURL, StringView(sample), sourcePosition, nullptr, URL(), nullptr, true);
```
Should just be able to pass `sample`, not `StringView(sample)`.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,42 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+
+            String sourceURL;
+            TextPosition sourcePosition(OrdinalNumber::beforeFirst(), OrdinalNumber());
+            String sample = makeString(sink, "|"_s, StringView(source).left(40));
```
Should use `'|'` instead of `"|"_s`.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,42 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+
+            const ContentSecurityPolicyDirectiveList& list = *policy.get();
+
+            String sourceURL;
```
Seems a little strange to use a local variable for this. Just pass `nullString()` instead?

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -776,6 +776,42 @@ AllowTrustedTypePolicy ContentSecurityPolicy::allowTrustedTypesPolicy(const Stri
+            String consoleMessage = makeString(policy->isReportOnly() ? "[Report Only] " : "",
+                "This requires a "_s, stringContext, " value else it violates the following Content Security Policy directive: \"require-trusted-types-for 'script'\""_s);
+
+            const ContentSecurityPolicyDirectiveList& list = *policy.get();
```
Should use `auto&` or just pass `*policy` below; no real value in putting it into a local variable.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -825,7 +861,7 @@ void ContentSecurityPolicy::reportViolation(const ContentSecurityPolicyDirective
 }
 
-void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirective, const ContentSecurityPolicyDirectiveList& violatedDirectiveList, const String& blockedURLString, const String& consoleMessage, const String& sourceURL, const StringView& sourceContent, const TextPosition& sourcePosition, JSC::JSGlobalObject* state, const URL& preRedirectURL, Element* element) const
+void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirective, const ContentSecurityPolicyDirectiveList& violatedDirectiveList, const String& blockedURLString, const String& consoleMessage, const String& sourceURL, const StringView& sourceContent, const TextPosition& sourcePosition, JSC::JSGlobalObject* state, const URL& preRedirectURL, Element* element, bool samplePreClipp ...
```
Using `bool` is not great for `samplePreClipped`. Just makes call sites very mysterious.

darinadler commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -848,7 +884,9 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
-    info.sample = violatedDirectiveList.shouldReportSample(effectiveViolatedDirective) ? sourceContent.left(40).toString() : emptyString();
+    info.sample = emptyString();
+    if (violatedDirectiveList.shouldReportSample(effectiveViolatedDirective))
+        info.sample = samplePreClipped ? sourceContent.toString() : sourceContent.left(40).toString();
```
This is not necessary. If the source content is already short, then it’s harmless to do the `left(40)` on it. If you want to bump this number up from 40 to something a little higher I guess that’s OK. Please don’t add `samplePreClipped`. I guess the tests are maybe written expecting a certain very specific type of truncation, but seriously this is too messy.

lukewarlow replied:
It's very specifically specced in this way currently.

I'm happy to tidy it up and just accept the test failures to keep this cleaner.

darinadler commented on Source/WebCore/dom/TrustedTypeUtils.cpp:
```diff
@@ -0,0 +1,162 @@
+    if (policyValue.isNull())
+        return nullptr;
+
+    if (expectedType == TrustedType::TrustedHTML)
```
This should be a `switch` statement, not a cascade of `if` statements.

darinadler commented on Source/WebCore/dom/TrustedTypePolicy.h:
```diff
@@ -49,6 +50,7 @@ class TrustedTypePolicy : public ScriptWrappable, public RefCounted<TrustedTypeP
     ExceptionOr<Ref<TrustedHTML>> createHTML(const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&&);
     ExceptionOr<Ref<TrustedScript>> createScript(const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&&);
     ExceptionOr<Ref<TrustedScriptURL>> createScriptURL(const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&&);
+    ExceptionOr<String> getPolicyValue(const TrustedType trustedTypeName, const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&&, bool throwIfMissing = true);
```
Should just be `TrustedType` not `const TrustedType`.

We try to avoid mystery `bool` arguments like this `throwIfMissing` one. Can we find a way to not have that, perhaps by having more than one function? I think one function can return null string if missing and the one that throws if missing can just call the other and post process it.

lukewarlow replied:
The problem with returning nullString and post processing is that the callbacks themselves can return null and they shouldn't throw in this situation.

I'll make it a boolean enum instead so it's more explicit.

darinadler replied:
I think you can split this into two separate functions. Whether internally you want to use a `bool` argument to a common function or can find a way to have one call the other is something we can tinker with. I am almost certain you can have one call the other but I can propose a specific way once I see your revised patch. For example if you can distinguish the specific exception you can convert the exception to a null string.

lukewarlow replied:
Doing something like in a GetPolicyValueForProcessDefault function could work? But it doesn't feel great, unfortunately the ExceptionCode alone isn't enough to distinguish because TypeErrors are thrown in other cases (such as from userland code)

```c++
    auto policyValue = getPolicyValue(trustedTypeName, input, WTFMove(arguments));
    if (policyValue.hasException()) {
        if (policyValue.exception().message().contains("TrustedTypePolicyOptions did not specify"_s))
            return String(nullString());
    }

    return policyValue;
```

darinadler replied:
Yes clearly wrong if it needs to check the string.

lukewarlow replied:
I've changed it to an enum so it's at least not a magic bool.

darinadler commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +51,68 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+        return Exception {
+            ExceptionCode::TypeError,
+            makeString("Policy "_s, m_name,
+            "'s TrustedTypePolicyOptions did not specify a create'"_s, trustedTypeToString(trustedTypeName).substring(7), "' member."_s)
```
This is not the traditional way to indent.

Calling `substring(7)` on the result of `trustedTypeToString` is an ugly way to handle it; we can do better. Maybe more than one function that returns a string for a `TrustedType`.

darinadler commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +51,68 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-policy-value-algorithm
+ExceptionOr<String> TrustedTypePolicy::getPolicyValue(const TrustedType trustedTypeName, const String& input, FixedVector<JSC::Strong<JSC::Unknown>>&& arguments, bool throwIfMissing)
```
Just `TrustedType` not `const TrustedType`.

darinadler commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -50,65 +51,68 @@ TrustedTypePolicy::TrustedTypePolicy(const String& name, const TrustedTypePolicy
+        ASSERT_NOT_REACHED();
+        return Exception {
+            ExceptionCode::TypeError
+        };
```
This should be a one-liner.

darinadler reviewed (changes requested):
Did a first pass and found some things to improve.

darinadler commented on Source/WebCore/bindings/js/JSDOMConvertStrings.h:
```diff
@@ -28,16 +28,22 @@
 WEBCORE_EXPORT String identifierToUSVString(JSC::JSGlobalObject&, const JSC::Identifier&);
 WEBCORE_EXPORT String valueToUSVString(JSC::JSGlobalObject&, JSC::JSValue);
 WEBCORE_EXPORT AtomString valueToUSVAtomString(JSC::JSGlobalObject&, JSC::JSValue);
+String getTrustedTypeCompliantString(TrustedType, JSC::JSGlobalObject&, JSC::JSValue, const String& sink, ShouldConvertNullToEmptyString);
```
WebKit coding style names this function without the word “get”.

lukewarlow replied:
Ah okay, fixed that.

lukewarlow committed:
Implement enforcement of `require-trusted-types-for` CSP directive
https://bugs.webkit.org/show_bug.cgi?id=267685

Reviewed by Darin Adler.

This patch implements the StringContext idl attribute to check
the `require-trusted-types-for` CSP and enforce trusted types accordingly.

This patch also makes use of the StringContext IDL attribute on an initial set of sinks.

More complicated sinks such as setAttribute, execCommand, eval and timer functions will be addressed in follow ups.

Spec: https://w3c.github.io/trusted-types/dist/spec/#integrations

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/reporting/report-clips-sample.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-source-file-path-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/IDLTypes.h:
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
(WebCore::trustedTypeCompliantString):
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
(WebCore::Converter<IDLStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::Converter<IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLLegacyNullToEmptyStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedHTMLAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedScriptAdaptor<T>>::convert):
(WebCore::Converter<IDLAtomStringStringContextTrustedScriptURLAdaptor<T>>::convert):
(WebCore::JSConverter<IDLAtomStringStringContextTrustedScriptURLAdaptor<T>>::convert):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateParametersCheck):
(IsAnnotatedType):
(GetAnnotatedIDLType):
(JSValueToNative):
* Source/WebCore/bindings/scripts/IDLAttributes.json:
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestStringContextConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.cpp: Added.
(WebCore::JSTestStringContextDOMConstructor::prototypeForStructure):
(WebCore::JSTestStringContextDOMConstructor::initializeProperties):
(WebCore::JSTestStringContextPrototype::finishCreation):
(WebCore::JSTestStringContext::JSTestStringContext):
(WebCore::JSTestStringContext::createPrototype):
(WebCore::JSTestStringContext::prototype):
(WebCore::JSTestStringContext::getConstructor):
(WebCore::JSTestStringContext::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLSetter):
(WebCore::JSC_DEFINE_CUSTOM_SETTER):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedHTMLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringGetter):
(WebCore::setJSTestStringContext_attributeWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedHTMLSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptSetter):
(WebCore::jsTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLGetter):
(WebCore::setJSTestStringContext_reflectedUrlAttributeWithStringContextTrustedScriptURLSetter):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLBody):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedHTMLAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptAndLegacyNullToEmptyStringBody):
(WebCore::jsTestStringContextPrototypeFunction_methodWithStringContextTrustedScriptURLAndLegacyNullToEmptyStringBody):
(WebCore::JSTestStringContext::subspaceForImpl):
(WebCore::JSTestStringContext::analyzeHeap):
(WebCore::JSTestStringContextOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestStringContextOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestStringContext::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestStringContext.h: Added.
(WebCore::JSTestStringContext::create):
(WebCore::JSTestStringContext::createStructure):
(WebCore::JSTestStringContext::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestStringContext.idl: Added.
* Source/WebCore/dom/Document+HTML.idl:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element+DOMParsing.idl:
* Source/WebCore/dom/InnerHTML.idl:
* Source/WebCore/dom/Range+DOMParsing.idl:
* Source/WebCore/dom/TrustedType.cpp: Added.
(WebCore::TrustedTypeVisitor::operator()):
(WebCore::trustedTypeToString):
(WebCore::trustedTypeToCallbackName):
(WebCore::processValueWithDefaultPolicy):
(WebCore::trustedTypeCompliantString):
* Source/WebCore/dom/TrustedType.h: Copied from Source/WebCore/workers/Worker.idl.
* Source/WebCore/dom/TrustedTypePolicy.cpp:
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
(WebCore::TrustedTypePolicy::getPolicyValue):
* Source/WebCore/dom/TrustedTypePolicy.h:
* Source/WebCore/html/HTMLEmbedElement.idl:
* Source/WebCore/html/HTMLIFrameElement.idl:
* Source/WebCore/html/HTMLObjectElement.idl:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::requireTrustedTypesForSinkGroup const):
(WebCore::ContentSecurityPolicy::allowMissingTrustedTypesForSinkGroup const):
(WebCore::ContentSecurityPolicy::reportViolation const):
* Source/WebCore/page/csp/ContentSecurityPolicy.h:
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.cpp:
(WebCore::ContentSecurityPolicyDirectiveList::shouldReportSample const):
* Source/WebCore/page/csp/ContentSecurityPolicyDirectiveList.h:
(WebCore::ContentSecurityPolicyDirectiveList::requiresTrustedTypesForScript const):
* Source/WebCore/workers/Worker.idl:
* Source/WebCore/workers/WorkerGlobalScope.idl:
* Source/WebCore/workers/service/ServiceWorkerContainer.idl:
* Source/WebCore/workers/shared/SharedWorker.idl:
* Source/WebCore/xml/DOMParser.idl:

Canonical link: https://commits.webkit.org/275607@main
